### ISSUE TITLE
getPath for shared data directory

### ISSUE THREAD

**Issue by TanninOne** [electron/electron#9926]
Would it be possible to add a constant to getPath that returns the path for a data directory that is shared between user accounts?
On windows this would be something like %ProgramData%/<appname> (CSIDL_COMMON_APPDATA)
On linux /var/lib/<appname>.
On mac /Library/Application Support/<appname>

**Comment by me0262** [comment]
Seconded.
Actually on Linux, shared application data is in /usr/share/<appname>

**Comment by tiangolo** [comment]
I would like that too!

---

For Linux I think it would be inside `/var/lib/`:

> An application (or a group of inter-related applications) must use a subdirectory of /var/lib for its data.

Reference: http://www.pathname.com/fhs/2.2/fhs-5.8.html

The `/usr/share` could be mounted on a network file system and shared by several systems:

> The "share" word is used because what is under /usr/share is platform independent, and can be shared among several machines across a network filesystem.

And `/var/lib` won't be modified by installing and un-installing packages:

> The name comes from "variable", because everything that is under this directory changes frequently, and the package system (RPM) doesn't keep control of. 

Reference: http://www.tldp.org/HOWTO/HighQuality-Apps-HOWTO/fhs.html

---

For Linux and macOS I guess the path is kind of fixed, for Windows I solved it with:

```JavaScript
// TODO generalize this to Linux and macOS
if (process.platform === 'win32') {
  const globalStoreDirPath = path.resolve(process.env.ProgramData || 'C:\\ProgramData', electron.app.getName());
  if (fs.existsSync(globalStoreDirPath)) {
    // Do stuff in that directory
  }
}
```

**Comment by codebytere** [comment]
working on this now

**Comment by poiru** [comment]
What is the use-case for this? None of these paths are writable by normal users (well, except `%ProgramData%` on Windows, but you need to set a proper ACL to make the files/directories you create writable by other users) so I'm not sure what Electron apps would use this for. Why can't the shared data just live in the app directory?

**Comment by TanninOne** [comment]
ProgramData is where applications on Windows are supposed to hold shared data/configuration files (Configuration data files the application uses and is per machine. It is used across all users of the application).
Do we really need a debate about if and why electron applications should be enabled to follow platform standards?

**Comment by poiru** [comment]
@TanninOne Yes, but the equivalent directories are not writable by normal users on other platforms. Why do we need a cross-platform API for a platform-specific problem?

**Comment by TanninOne** [comment]
how is "find location where data shared between users should be stored" a platform-specific problem?
On what platform(s) are you working where this is never required?

**Comment by codebytere** [comment]
well, i'm verifying paths rn but the infrastructure for the feature is in the PR so feel free to migrate comments over there @TanninOne @poiru

**Comment by poiru** [comment]
@TanninOne Well, on macOS, most Electron apps are distributed as bundles that can be installed without write access to `/Library`.

I'm curious what hear about your actual use-case. For your app, what or who creates the files in the shared directory?

**Comment by TanninOne** [comment]
Any user on the local system really. My application lets the user choose if he wants to store his configuration just for himself or with other users on the same PC.

**Comment by poiru** [comment]
Hmm, interesting. I guess my concern with this API is that you will still need to have platform-specific code for macOS/Linux. Code that works on Windows will not necessary work on macOS/Linux because the directories are unwritable for normal, unelevated processes. I guess that could be "addressed" with proper documentation.

**Comment by TanninOne** [comment]
But that's nothing new, the electron api is full of features that work only on one or two of the available platforms or work differently between them. And what would be the alternative? Offer only apis for features that work exactly the same across all platform?

**Comment by MarshallOfSound** [comment]
> But that's nothing new, the electron api is full of features that work only on one or two of the available platforms or work differently between them. And what would be the alternative? Offer only apis for features that work exactly the same across all platform?

The case here is not an API that only works on some platforms, or works differently on some platforms.  What you're suggesting is an API that would be technically unusable for most users.  Most Electron apps install themselves with user level privileges, this means this path would be inaccessible most of the time.

I'm not saying this API shouldn't happen, I'm just saying it's distinctly different to other API's we've made in the past.  It's also not even that on some devices it won't be accessible, it's also that even if they can access it if you don't write the files with the correct permissions other users won't be able to access them.  (looking at you windows)

TLDR: I'm slightly more in favor of this being a userland concept where the developer would have to be more aware of what they are doing and what it means rather than blanket offer this as another path that Electron apps can use

**Comment by tiangolo** [comment]
Here are my 2 cents:

We are building an app that we want to be installable on many users' computers (a corporate environment). We want to be able to have user-specific settings in the normal `app.getPath('userData')` directory (`%APPDATA%` in Windows).

But we also want to be able to set configurations for that organization that apply to all the installations, independent of the user. 

We want those configurations to be independent of each organization using the app. So, we cannot hardcode them.

But we want those configurations to be read and used by the app for all the users in that organization.

In the case of Windows, when they use those "corporate" Windows Domains and stuff, a specific user might not exist on a specific machine. But once he tries to log in, the machine communicates to the "Windows Domain server" (or something like that), checks the user's credentials and then creates the user. Because of that, we cannot just pre-populate every user's `userData` directory with our configurations, because the user doesn't exist yet by the time we configure, install and set up everything.

I agree that writing to that location from the app would not be very feasible in many environments due to permissions, but at least in our case, we need only read permissions. And we actually need to be able to read things that were there before the installation of the Electron app. It would be "our problem" to set the permissions right for Linux and macOS users for those global app location configurations.

But it would be cool to be able to have another option for `app.getPath()` to retrieve that directory independent of the platform. In our case, we target the 3 OSes (I'm actually a Linux guy), so we end up doing that with custom code like in my comment above: https://github.com/electron/electron/issues/9926#issuecomment-321299417

**Comment by codebytere** [comment]
@tiangolo for linux `/var/lib` is unwritable, and so is `ProgramData` on win. I'd like to make this happen, but unless there's a clear path forward for at least 2 of the three platforms i might say it's better off in userland too

**Comment by tiangolo** [comment]
*TL;DR:*

The only reason to implement this would be if having a pre-defined "Electron standard" read-only global directory (e.g. for pre-made configs not created by the app) is a valuable feature.

I know it's a use case a bit contrived. So, if that specific use case wouldn't be something valuable to have as part of Electron, you shouldn't implement it.

I can only imagine how that would be useful when installing an application for a large group of users (like a corporate environment). And have configurations that apply to all those users in that specific group, and have those group specific settings apply without needing any other interaction from the users. That's our specific use case.

---

Yep @codebytere , `/var/lib` is unwritable in Linux, permissions would have to be set by hand beforehand.

I'm pretty sure that for macOS it would be very similar to Linux.

For `ProgramData` in Windows, [I understand it's actually writable by anyone](https://docs.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-shell-setup-folderlocations-programdata) (I don't know why, Windoze, I guess).

> [...] this folder can be used by applications to store data for standard users, because it does not require elevated permissions.

Still, for cross-compatibility, I would say that it's at most a "read only path". And I bet some developers reading the API docs would ask themselves why Electron would support such a thing.

---

I don't really know of a path in any of the 3 platforms that is writable and global (in contrast to user-specific).

I guess the only use case for this would be to be able to preset system-wide read-only data by other mechanisms, not by the Electron app installation, but that would be read by the Electron app.

I know that all the platforms have writable global temporal directories, but those directories don't have any persistence guarantee, so it wouldn't fit for that specific use case of a global config data directory.

I see that, for Windows, it seems to be that `ProgramData` directory, named "Common Application Data Folder", [here](https://docs.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-shell-setup-folderlocations-programdata) and [here](https://msdn.microsoft.com/en-us/library/windows/apps/hh465094.aspx).

That's what I see in the (I would say) scarce, incomplete and confusing Windows documentation :disappointed: . So, I'm never sure if I'm coding things right when interacting with Windows, not even when I had to read a lot of those Windows docs to [use them with `node-ffi`](https://github.com/sindresorhus/active-win/pull/7).

I understand if you decide that this shouldn't be part of Electron itself and should be done in "userland". 

That would make it clear that this won't be part of Electron officially and would make this feature / hack a good candidate for a third party package based on something like my snippet above. 

Still, knowing there's a decision from you (the Electron team) about it, would help to define if it's worthwhile (or not) to write an independent package to try and solve that specific use case. :smiley:

**Comment by poiru** [comment]
@tiangolo Even on Windows, you need to manually set the proper ACL to make folders/files under `ProgramData` writable for other users. See: https://stackoverflow.com/questions/22107812/privileges-owner-issue-when-writing-in-c-programdata

**Comment by tiangolo** [comment]
Ah! I see @poiru . Thanks for the clarification.

Then, in all the cases, it would have to be documented as a read-only directory. For all the platforms.

**Comment by MarshallOfSound** [comment]
> Then, in all the cases, it would have to be documented as a read-only directory. For all the platforms.

Doesn't this make it a 99% useless folder?  Even if we document it strictly as read only I can see the issues already "Can't write to 'sharedData' path, plz help".  I'd prefer to leave this out of core and let developers make the decision to read stuff from these folders if they so choose.

**Comment by TanninOne** [comment]
I guess the purpose of getPath is different from what I thought it was. My assumption was that getPath is about enabling applications to obey OS standards, i.e. "getPath('temp')" is asking electron "on the current OS, where am I **supposed** to store temporary data".
Under this assumption a "on the current OS, where am I supposed to store generated files shared with other users" makes sense, independent of access rights, because it's not about how a directory works but about what it's purpose is.

But it seems, from this discussion, it's more about the _how_: "give me a writable location that gets deleted occasionally" or "give me some writable location that the current user can write to but that is not accessible to other users".

Which is inconsequent imho because if I'm not mistaken some of the paths returned by getPath already have different functionality, i.e. the path returned by getPath('temp') on windows is accessible only to the current user whereas on linux it probably returns /tmp which is shared between all accounts. And constants like "pictures", "videos", ... clearly sound like they are about purpose not functionality, I would claim it doesn't make a difference if on windows the "videos" directory was readable by other users and on linux it wasn't, as long as the OS wants users to store videos there.

**Comment by TanninOne** [comment]
> Doesn't this make it a 99% useless folder?

How
does that make the path useless?
What's stopping a user from creating a subdirectory with the appropriate access rights? Like the OS vendor intended? Like countless other applications have done before?

**Comment by codebytere** [comment]
As we strive to keep the Electron API itself thin, this is not something that we think should be added into core functionality and instead leave it to developers to make the decision to read and write to this folder if they so choose, as @MarshallOfSound said above.

**Comment by tiangolo** [comment]
OK, thanks @codebytere for coming back and letting us know the decisions made. So, it's on our side (the developers) now. :+1: :smile:

**Comment by KaminoRyo** [comment]
I have same problem too.
For my app, Only the administrator can read and write config file, and general users can read.
For macOS, I decided to use the "Shared" directory. 
Because writing to the top level "Library/Application Support" requires authorization and is difficult to program. 😔
